### PR TITLE
Column Chart API update

### PR DESCRIPTION
**Pull request by jdelight**
Update to column charts to use the new declarative API.

- [x] Move Bar logic to ColumnChart
- [x] Add BarChart Context to ColumnChart
- [x] Create new MirroredColumnChart component
- [x] Update tests

### CODE REVIEW

File: src/charts/column-chart/ColumnChart.js [hunk 1]
```diff
@@ -2,37 +2,121 @@ import PropTypes from 'prop-types';
 import React, { Component } from 'react';
 import omit from 'lodash.omit';
 import { Base } from 'bw-axiom';
+import Bars from '../bars/Bars';
+import BarChartContext from '../bar-chart/BarChartContext';
```
Comment (HHogg) [inline src/charts/column-chart/ColumnChart.js]: Good shout on reusing this 👍

File: src/charts/column-chart/ColumnChart.js [hunk 2]
```diff
@@ -2,37 +2,121 @@ import PropTypes from 'prop-types';
...
+import ColumnChartVisual from './ColumnChartVisual';
+import ColumnChartBars from './ColumnChartBars';
 import './ColumnChart.css';
+import { formatData } from './utils';
 
 export default class ColumnChart extends Component {
   static propTypes = {
-    children: PropTypes.node.isRequired,
+    ContextComponent: PropTypes.func,
+    Label: PropTypes.func,
```
Comment (HHogg) [inline src/charts/column-chart/ColumnChart.js]: I think this should probably be a pre-rendered label `label: PropTypes.node`. It differs slightly from `ContextComponent`, because `ContextComponent` has to be dynamic depending on the Bars/DotPlots it is wrapping. 

_See comment [1] below_

File: src/charts/column-chart/ColumnChart.js [hunk 3]
```diff
@@ -2,37 +2,121 @@ import PropTypes from 'prop-types';
...
+    chartKey: PropTypes.arrayOf(PropTypes.shape({
+      color: PropTypes.oneOf([
```
Comment (HHogg) [inline src/charts/column-chart/ColumnChart.js]: Really minor, we realised when doing the DotPlot and BarChart API changes that this color property would be checked at the very end of the chain when it reaches `<Bar>`. So here could be changed to just `PropTypes.string.isRequired`

File: src/charts/column-chart/ColumnChart.js [hunk 4]
```diff
@@ -2,37 +2,121 @@ import PropTypes from 'prop-types';
...
+    direction: PropTypes.string,
     height: PropTypes.string.isRequired,
+    labelColumnWidth: PropTypes.string,
+    showLabel: PropTypes.bool,
+    showSubLabel: PropTypes.bool,
+    showXAxisLabels: PropTypes.bool,
     yAxisWidth: PropTypes.string,
   };
 
   static childContextTypes = {
```
Comment (HHogg) [inline src/charts/column-chart/ColumnChart.js]: Now that it's fully encapsulated you can ditch all of the context behaviour and just pass the props down through the components :) 

_See comment [1] below_

File: src/charts/column-chart/ColumnChart.js [hunk 5]
```diff
@@ -2,37 +2,121 @@ import PropTypes from 'prop-types';
...
+      label: PropTypes.node.isRequired,
+      values: PropTypes.object.isRequired,
+    })).isRequired,
+    direction: PropTypes.string,
     height: PropTypes.string.isRequired,
+    labelColumnWidth: PropTypes.string,
+    showLabel: PropTypes.bool,
+    showSubLabel: PropTypes.bool,
+    showXAxisLabels: PropTypes.bool,
     yAxisWidth: PropTypes.string,
```
Comment (HHogg) [inline src/charts/column-chart/ColumnChart.js]: Is this what has become `labelColumnWidth`?

File: src/charts/column-chart/ColumnChart.js [hunk 6]
```diff
@@ -2,37 +2,121 @@ import PropTypes from 'prop-types';
...
 
     return (
-      <Base { ...omit(rest, ['yAxisWidth']) }
+      <Base { ...omit(rest, ['labelColumnWidth', 'yAxisWidth']) }
           className="ax-column-chart"
           style={ { height } }>
-        { children }
+        <ColumnChartRow>
+          { Label &&
+            <Label align={ this.getLabelAlignment() }></Label>
```
Comment (HHogg) [inline src/charts/column-chart/ColumnChart.js]: __[1]__ `<ColumnChartYAxis>` is quite essential to the layout, so relating to the two comments above, if the pre-rendered label given was placed inside like ...

```jsx
{ label && (
  <ColumnChartYAxis>
    { label }
  </ColumnChartYAxis>
) }
```

... it would allow you to remove the context and enforce the layout.

File: src/charts/column-chart/ColumnChart.test.js [hunk 7]
```diff
@@ -4,7 +4,10 @@ import { ColumnChart } from 'bw-axiom';
 
 function getComponent(props = {}) {
   return renderer.create(
-    <ColumnChart height="10rem" { ...props }>
+    <ColumnChart
```
Comment (HHogg) [inline src/charts/column-chart/ColumnChart.test.js]: Similar to with the DotPlotChart and BarChart, these tests have become less helpful because of their size. It should be sufficient enough to test the utils.

File: src/charts/column-chart/ColumnChartXAxis.test.js [hunk 8]
```diff
@@ -4,7 +4,7 @@ import { ColumnChart, ColumnChartXAxis } from 'bw-axiom';
 
 function getComponent(props = {}) {
   return renderer.create(
-    <ColumnChart yAxisWidth="10rem">
+    <ColumnChart chartKey={ [{}] } data={ [{ label: 'test', values: [] }] } yAxisWidth="10rem">
```
Comment (HHogg) [inline src/charts/column-chart/ColumnChartXAxis.test.js]: `yAxisWidth` is now `labelColumnWidth`?

File: src/charts/column-chart/MirroredColumnChart.js [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+import ColumnChartXAxis from './ColumnChartXAxis';
+import ColumnChartXAxisLabel from './ColumnChartXAxisLabel';
+import './ColumnChart.css';
+import { formatData } from './utils';
+
+export default class MirroredColumnChart extends Component {
+  static propTypes = {
+    ContextComponent: PropTypes.func,
+    Label: PropTypes.func,
+    ReflectionLabel: PropTypes.func,
```
Comment (HHogg) [inline src/charts/column-chart/MirroredColumnChart.js]: Similar to comment above with the pre-rendered `Label` and `ReflectionLabel`

File: src/charts/column-chart/MirroredColumnChart.js [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+    chartKey: PropTypes.arrayOf(PropTypes.shape({
+      color: PropTypes.oneOf([
```
Comment (HHogg) [inline src/charts/column-chart/MirroredColumnChart.js]: Similar comment above with skipping the strict check here.

File: src/charts/column-chart/MirroredColumnChart.js [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+    height: PropTypes.string.isRequired,
+    labelColumnWidth: PropTypes.string,
+    reflectionData: PropTypes.arrayOf(PropTypes.shape({
+      label: PropTypes.node.isRequired,
+      values: PropTypes.object.isRequired,
+    })).isRequired,
+    showLabel: PropTypes.bool,
+  };
+
+  static childContextTypes = {
```
Comment (HHogg) [inline src/charts/column-chart/MirroredColumnChart.js]: Similar to comment above about ditching the context.

File: src/charts/column-chart/MirroredColumnChart.js [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+      Label,
+      ReflectionLabel,
+      chartKey,
+      data,
+      height,
+      reflectionData,
+      ...rest
+    } = this.props;
+
+    const formattedData = formatData(chartKey, data);
```
Comment (HHogg) [inline src/charts/column-chart/MirroredColumnChart.js]: Do you need to format the data here?

File: src/index.js [hunk 13]
```diff
@@ -120,6 +120,7 @@ export { default as ColumnChartRow } from './charts/column-chart/ColumnChartRow'
 export { default as ColumnChartVisual } from './charts/column-chart/ColumnChartVisual';
 export { default as ColumnChartXAxis } from './charts/column-chart/ColumnChartXAxis';
 export { default as ColumnChartXAxisLabel } from './charts/column-chart/ColumnChartXAxisLabel';
```
Comment (HHogg) [inline src/index.js]: All of the ColumnChart component scan also be removed from the export :)

File: src/charts/column-chart/ColumnChart.js (hunk 6 above)
Comment (lpoulter) [inline src/charts/column-chart/ColumnChart.js]: How would i prerender the label? Would i use the children property?

File: src/charts/bar-chart/BarChart.js [hunk 14]
```diff
@@ -29,6 +29,7 @@ export default class BarChart extends Component {
     })).isRequired,
     expandButtonSuffix: PropTypes.string,
     labelColumnWidth: PropTypes.string.isRequired,
+    showBarLabel: PropTypes.bool,
```
Comment (HHogg) [inline src/charts/bar-chart/BarChart.js]: For clarity in the release notes this should really be split out into a separate commit... 

```
feat(BarChart): Added `showBarLabel` option to conditionally show bar labels
```

... kind of thing.

File: src/charts/bar-chart/BarChart.js [hunk 15]
```diff
@@ -95,12 +97,13 @@ export default class BarChart extends Component {
                         color={ color }
                         data={ data[index] }
                         dataIndex={ index }
+                        hoverEffect={ !showBarLabel }
```
Comment (HHogg) [inline src/charts/bar-chart/BarChart.js]: This `hoverEffect` prop could be avoided by conditionally setting the onMouseEnter/Leave props ... 

```jsx
onMouseEnter={ showBarLabel && this.onMouseEnter }
onMouseLeave={ showBarLabel && this.onMouseLeave }
```

And likewise inside the `BarChartContext`

```jsx
onMouseEnter={ onMouseEnter && () => onMouseEnter(dataIndex, color) }
onMouseLeave={ onMouseLeave }
```

As it is `BarChart` that controls the hover behaviour the logic should probably remain here.

File: src/charts/column-chart/ColumnChart.js [hunk 16]
```diff
@@ -2,37 +2,129 @@ import PropTypes from 'prop-types';
...
+    labelColumnWidth: PropTypes.string,
+    showBarLabel: PropTypes.bool,
+    showSubLabel: PropTypes.bool,
+    showXAxisLabels: PropTypes.bool,
     yAxisWidth: PropTypes.string,
   };
 
   static defaultProps = {
+    labelColumnWidth: '0rem',
     yAxisWidth: '0rem',
```
Comment (HHogg) [inline src/charts/column-chart/ColumnChart.js]: This default can be removed now.

File: src/charts/column-chart/ColumnChart.js [hunk 17]
```diff
@@ -2,37 +2,129 @@ import PropTypes from 'prop-types';
...
+                  { values.map(({ color, value }, index) =>
+                    <BarChartContext
+                        ContextComponent={ ContextComponent }
+                        color={ color }
+                        data={ data[index] }
+                        dataIndex={ index }
+                        hoverEffect={ !showBarLabel }
+                        key={ color }
+                        label={ label }
+                        labelStrong={ index === this.state.hoverIndex }
```
Comment (HHogg) [inline src/charts/column-chart/ColumnChart.js]: I think there is an issue with the logic here, only the value of the bar you are hovering on should have strong and stable styling (see DotPlot and BarChart). Currently all of the values for that data set are strong.

File: src/charts/column-chart/ColumnChart.js [hunk 18]
```diff
@@ -2,37 +2,119 @@ import PropTypes from 'prop-types';
...
-    yAxisWidth: PropTypes.string,
-  };
-
-  static childContextTypes = {
+    label: PropTypes.node,
+    labelColumnWidth: PropTypes.string,
+    showBarLabel: PropTypes.bool,
+    showSubLabel: PropTypes.bool,
+    showXAxisLabels: PropTypes.bool,
     yAxisWidth: PropTypes.string,
```
Comment (HHogg) [inline src/charts/column-chart/ColumnChart.js]: There's still a bunch of references to `yAxisWidth` ... 

https://github.com/BrandwatchLtd/axiom/pull/235/files#diff-f509cc4d63cd94d99cb1e64bfd798f25R34
https://github.com/BrandwatchLtd/axiom/pull/235/files#diff-f509cc4d63cd94d99cb1e64bfd798f25R66
https://github.com/BrandwatchLtd/axiom/pull/235/files#diff-f509cc4d63cd94d99cb1e64bfd798f25R80
https://github.com/BrandwatchLtd/axiom/pull/235/files#diff-f509cc4d63cd94d99cb1e64bfd798f25R85
https://github.com/BrandwatchLtd/axiom/pull/235/files#diff-f509cc4d63cd94d99cb1e64bfd798f25R111
https://github.com/BrandwatchLtd/axiom/pull/235/files#diff-42f3c16081cd0d05cdc86dcdb7dd1c97R11
https://github.com/BrandwatchLtd/axiom/pull/235/files#diff-42f3c16081cd0d05cdc86dcdb7dd1c97R19
https://github.com/BrandwatchLtd/axiom/pull/235/files#diff-90d1b43857d856d643b5b67f2e52e805R7
https://github.com/BrandwatchLtd/axiom/pull/235/files#diff-07f59ede855602d6184a5202c1eac5d8R50
https://github.com/BrandwatchLtd/axiom/pull/235/files#diff-a2322b55a05fa2cfd7095302a647577fR13
https://github.com/BrandwatchLtd/axiom/pull/235/files#diff-55a4241b2c821611a3622d3cff432d5cR31

File: src/charts/column-chart/ColumnChart.js [hunk 19]
```diff
@@ -2,37 +2,119 @@ import PropTypes from 'prop-types';
...
+    const hoverProps = {};
+
+    if (!showBarLabel) {
+      hoverProps.onMouseEnter = this.onMouseEnter;
+      hoverProps.onMouseLeave = this.onMouseLeave;
+    }
 
     return (
-      <Base { ...omit(rest, ['yAxisWidth']) }
+      <Base { ...omit(rest, ['labelColumnWidth', 'yAxisWidth']) }
```
Comment (HHogg) [inline src/charts/column-chart/ColumnChart.js]: This omit won't be needed once you remove `yAxisWidth`

File: src/charts/column-chart/ColumnChart.js [hunk 20]
```diff
@@ -2,37 +2,119 @@ import PropTypes from 'prop-types';
...
+            { label }
+          </ColumnChartYAxis>
+          }
+          <ColumnChartVisual>
+            { formattedData.map(({ values, label, subLabel }, index) =>
+              <ColumnChartBars key={ index }>
+                <Bars direction={ direction } label={ showSubLabel && subLabel }>
+                  { values.map(({ color, value }) =>
+                    <BarChartContext
+                        { ...hoverProps }
```
Comment (HHogg) [inline src/charts/column-chart/ColumnChart.js]: To avoid needing to move around the file you could conditionally set the props inline with the component.

File: src/charts/column-chart/ColumnChartRow.css [hunk 21]
```diff
@@ -4,7 +4,4 @@
   display: flex;
   flex: 1 1 0%;
   margin: calc(var(--spacing-grid-size) * 3) 0;
-
-  &:first-child { margin-top: 0; }
-  &:last-child  { margin-bottom: 0; }
```
Comment (HHogg) [inline src/charts/column-chart/ColumnChartRow.css]: By removing this, it's going to add unexpected margin to the top and bottom of the chart. 
screen shot 2017-06-19 at 10 37 11

It would be better to move the margin to the axis component and still have this `first-child/last-child` styling so it is removed from the top and bottom when not used in a mirrored scenario. 

screen shot 2017-06-19 at 10 37 25

File: src/charts/column-chart/MirroredColumnChart.js [hunk 22]
```diff
@@ -0,0 +1,82 @@
...
+      <Base { ...rest }
+          className="ax-column-chart"
+          style={ { height } }>
+        <ColumnChart
+            { ...rest }
+            ContextComponent={ ContextComponent }
+            chartKey={ chartKey }
+            data={ data }
+            direction="up"
+            height="20rem"
```
Comment (HHogg) [inline src/charts/column-chart/MirroredColumnChart.js]: These heights shouldn't be fixed and should com from the props.

File: src/charts/column-chart/MirroredColumnChart.js [hunk 23]
```diff
@@ -0,0 +1,82 @@
...
+      data,
+      height,
+      labelColumnWidth,
+      reflectionData,
+      ...rest
+    } = this.props;
+
+    return (
+      <Base { ...rest }
+          className="ax-column-chart"
```
Comment (HHogg) [inline src/charts/column-chart/MirroredColumnChart.js]: This class has already been used inside ColumnChart, perhaps this should be `ax-mirrored-column-chart`?

File: src/charts/column-chart/MirroredColumnChart.js [hunk 24]
```diff
@@ -0,0 +1,82 @@
...
+          ) }
+        </ColumnChartXAxis>
+
+        <ColumnChart
+            { ...rest }
+            ContextComponent={ ContextComponent }
+            chartKey={ chartKey }
+            data={ reflectionData }
+            direction="down"
+            height="20rem"
```
Comment (HHogg) [inline src/charts/column-chart/MirroredColumnChart.js]: Same fixed heights here.

File: src/charts/column-chart/MirroredColumnChart.js [hunk 25]
```diff
@@ -0,0 +1,82 @@
...
+      reflectionData,
+      ...rest
+    } = this.props;
+
+    return (
+      <Base { ...rest }
+          className="ax-column-chart"
+          style={ { height } }>
+        <ColumnChart
+            { ...rest }
```
Comment (HHogg) [inline src/charts/column-chart/MirroredColumnChart.js]: The rest of the props have been spread on multiple components, this should probably just be on the Base component at the top?

File: src/charts/column-chart/MirroredColumnChart.test.js [hunk 26]
```diff
@@ -0,0 +1,22 @@
...
+        chartKey={ [{}] }
+        data={ [{ label: 'test', values: [] }] }
+        height="20rem"
+        labelColumnWidth="10rem"
+        reflectionData={ [{ label: 'test', values: [] }] }/>
+  );
+}
+
+describe('MirroredColumnChart', () => {
+  it('renders with defaultProps', () => {
```
Comment (HHogg) [inline src/charts/column-chart/MirroredColumnChart.test.js]: Similar to ColumnChart.test you can ditch this one. They're just too big to be of any use.

File: src/charts/column-chart/ColumnChartXAxis.css [hunk 27]
```diff
@@ -1,4 +1,7 @@
+@import '../../materials/sizing';
+
 .ax-column-chart__x-axis {
   display: flex;
   flex: 0 0 auto;
+  margin: calc(var(--spacing-grid-size) * 3) 0;
```
Comment (HHogg) [inline src/charts/column-chart/ColumnChartXAxis.css]: The margins need removing from ColumnChartRows component because they are currently doubling up. 

And first-child/last-child rules need adding here to prevent adding margins where they are not needed.
